rfcs/0006-configuration-constants.md

```diff
@@ -0,0 +1,128 @@
+
+```rust
+// main.sw
+script;
+
+configurable {
+    CONTRACT_ADDRESS: b256,
+    PRICE_RATIO: u64,
+}
+
+fn main() { ... }
+```
+```
+// configurable-constants.json
+
+{
+    "CONTRACT_ADDRESS": {
+        "type": "b256",
+        "offset": "12345"
+    }
+}
+```
+
+If you are not using the SDK, it will be possible to define these values by specifying a `constants.sw` flag. This will work with the [notion of contract dependencies](https://github.com/FuelLabs/sway-rfcs/pull/19#issuecomment-1286351753). This Sway file must export the required constants as specified by the `configurable` block.
+
+
+# Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+The `type` field of the JSON file should describe the type of the configurable in the same manner as types are described in the JSON ABI file.
+
+No two configurables in a program should have the same name to prevent collisions.
+
+The range from the offset to the offset plus the size of the type will represent bytes within the bytecode that will be overwritten by the SDK, which will write values according to the ABI encoder's memory layout, which is consistent with Sway's memory layout.
+
+The change is not breaking as it is entirely new functionality.
+
+Configuration time constants may not be defined in any library code and must be defined at the top level. Should a library need to reference a configurable value, it should utilize design patterns such as the [builder pattern](https://en.wikipedia.org/wiki/Builder_pattern) to ingest the constants from the top level.
+
+The `pub` keyword will operate as normal: if it is `pub`, then it is public and importable from elsewhere in the Sway program. If not, then it is only referencable in the scope it is defined within.
+
+## Deploying with and without the SDK
+
+With the SDK, the deployment and assignment of values will all be handled within the SDK's API. Without the SDK, `forc` will accept command line arguments to specify these values, or allow for their specification in the manifest. In the future, we may wish to introduce more methods of defining these values, like environment variables, but it will not be necessary for the MVP. If the values are specified in more than one way, the SDK would take precedence. This is because the SDK does not actually compile the bytecode in this process; it merely overwrites the indices. This is final and precedence cannot be specified by the user.
+
+## Failure to provide constants
+
+If the constants are not provided, the SDK and/or `forc` should throw an error. Both tools have knowledge of the required constants due to the descriptor file, so they may check for the constants' presence and report a descriptive error if any are missing.
```

Comment:
By `forc` here we mean `forc deploy` specifically? I assume `forc build` shouldn't care as it should be able to emit bytecode with default initialized config constants, that are later specified by the SDK.

Reply:
`forc build` *does* care due to [the Contract Dependencies feature](https://fuellabs.github.io/sway/master/forc/manifest_reference.html#the-contract-dependencies-section). During `build`, we generate contract ID constants for each contract dependency, and the configurable value for each contract dependency must be known in order to produce the correct ID.

Reply:
To elaborate, this doesn't stop the user from updating `configurable` values later on using the SDK to create a contract factory or whatever.

It's just that we should expect that users **will** misuse the feature and expose configurable values from contracts that don't really need them to be configurable. As a result, we need to be able to specify these values when these contracts are specified within a forc project's `[contract-dependencies]` table, otherwise we can't guarantee a correct ID for those users who actually want a declarative approach to contract deployment.

---

W.r.t. this "Failure to provide constants" section, @mohammadfawaz I like your idea [here](https://github.com/FuelLabs/sway-rfcs/pull/19#discussion_r1003901756) of requiring that any declared `configurable` values always have some initial value (whether explicitly specified or defining some "zeroed" default for each type). That would remove need to throw an error at all I guess, as at the very least the configurable bytecode should always be valid?